Source fetching

Leonidas-from-XIV opened the pull request:
This is a PR to implement source fetching as suggested in #7158. It uses the opam-libraries to construct a command to download the source, which has the advantage of using the same fetching commands as OPAM does, but it wraps the execution using the process spawning/concurrency logic of Dune.

Requires https://github.com/ocaml/opam/pull/5452 to access the contents of the `command` type, to execute the command.

Currently work in progress as it still requires some testing and some parts are implemented in a very rudimentary way.

Leonidas-from-XIV committed:
Implement API to fetch sources using OPAM libraries

Signed-off-by: Marek Kubica <marek@tarides.com>

rgrinberg commented on src/dune_pkg/fetch.ml:
```diff
@@ -0,0 +1,73 @@
+      | None | Some false -> None
+    in
+    let stderr_to = Process.Io.stderr in
+    let stdout_to = Process.Io.stdout in
```
Is that dune's stdout/stderr? If so, that's definitely not allowed as it will ruin the display.

Leonidas-from-XIV replied:
This was just temporary to get something working. I've fixed it since.

rgrinberg commented on src/dune_pkg/fetch.ml:
```diff
@@ -0,0 +1,73 @@
+  let path = Path.to_string target in
+  let fname = OpamFilename.of_string path in
+  let hashes = [] in
+  let complete = Fiber.Ivar.create () in
```
Don't think you need this ivar. You're waiting for the subsequent bind to complete before filling it anyway

Leonidas-from-XIV replied:
Ah, indeed. It was a leftover from some previous code attempts of running the job within the OPAM machinery before implementing `Fiber_job`. Removed.

rgrinberg commented on src/dune_pkg/dune:
```diff
@@ -0,0 +1,6 @@
+(library
+ (name dune_pkg)
+ (synopsis "[Internal] Dune's packaging support")
+ (libraries stdune fiber dune_engine opam-core opam-repository opam-format)
```
Take a look at how we vendor other libraries. We'll need to do the same for these two.

Leonidas-from-XIV replied:
I've added an `update-opam.sh` and vendored a modified copy. Will need to wait for https://github.com/ocaml/opam/pull/5452 to be merged to update it to the "official" repo and I also cherry-picked the changes from https://github.com/ocaml/opam/pull/4771 to make opam build with `implicit_transitive_deps false`.

rgrinberg replied:
if you have any changes to vendored libraries, our convention is to fork them into opam-dune and keep the patches there.

Leonidas-from-XIV replied:
Ah, good, then I'll fork the stubbed-out opam (`OpamConsole`) to over there and update the script to pull it, at which point I can also remove the need for `configure.ac`.

Leonidas-from-XIV replied:
Done so now.

rgrinberg commented on src/dune_pkg/fetch.ml:
```diff
@@ -0,0 +1,73 @@
+module Process = Dune_engine.Process
+module Display = Dune_engine.Display
+
+module FiberJob = struct
```
Dune's convention is `Fiber_job` rather than `FiberJob`.

rgrinberg commented on src/dune_pkg/fetch.ml:
```diff
@@ -0,0 +1,93 @@
+let fetch url ~target =
+  let open Fiber.O in
+  let path = Path.to_string target in
+  Format.eprintf "Path to fetch to is %s\n" path;
```
We have a `Log` module in `Dune_util`

rgrinberg commented on test/unit-tests/source_fetch/source_fetch.ml:
```diff
@@ -0,0 +1,38 @@
+      ; signal_watcher = `No
+      }
+  in
+  let on_event _config event =
```
You can just make this callback do nothing. It's only useful for builds.

rgrinberg commented on test/unit-tests/source_fetch/source_fetch.ml:
```diff
@@ -0,0 +1,38 @@
+
+let main () =
+  let open Fiber.O in
+  let url = OpamUrl.of_string "https://xivilization.net/~marek/kimiko.txt" in
```
Can we make this test not depend the network? Also, I'd like to see some tests with git repos, tarballs, compressed tarballs, etc.

rgrinberg replied:
Yeah, a local http server is a decent option.

Leonidas-from-XIV replied:
Added the most basic local HTTP server that speaks just enough HTTP to serve one request and exit.

Leonidas-from-XIV commented on src/dune_pkg/fetch.ml:
```diff
@@ -0,0 +1,73 @@
+  let fname = OpamFilename.of_string path in
+  let hashes = [] in
+  let complete = Fiber.Ivar.create () in
+  let job = OpamRepository.pull_file path fname hashes [ url ] in
```
1. There are some calls to `OpamConsole.msg` but these are not printed in the default settings. I am not sure if this is a problem.
2. Yes, it will write to the OPAM cache. We have two options:
   1. Disable the cache altogether
   2. Specify our own cache dir, but if so, which directory to pick?

rgrinberg replied:
1. ok good. that's good enough for us.
2. not sure if sharing the cache is a good idea, but for now it doesn't matter too much. i'd just leave a comment about this so that we're aware.

Leonidas-from-XIV replied:
In any case, `OpamConsole` is now stubbed out in our fork so it will never print anything, no matter the OPAM configuration.

Leonidas-from-XIV commented on src/dune_pkg/fetch.ml:
```diff
@@ -0,0 +1,93 @@
+
+  let* downloaded = Fiber_job.run job in
+  match downloaded with
+  | Up_to_date () | Result () -> Fiber.return ()
```
`Up_to_date` is returned when it has found the file in the cache. I don't think `pull_file` actually ever returns this variant, as it handles `Up_to_date` internally and turns it into `Result`.

rgrinberg commented on vendor/opam/src/core/opamConsole.mli:
```diff
@@ -0,0 +1,176 @@
+(**************************************************************************)
```
We need to make sure that our code never exercises the code in this module. Perhaps we could replace it with a stub?

Leonidas-from-XIV replied:
I've stubbed it out and added a line to the update script to load the stubbed version.

Leonidas-from-XIV commented:
The alternative is to implement the same thing on our own (I think OPAM also does some wrangling with `tar`, `gzip`, `bzip`, `git` etc) which would reduce the amount of code added but the advantage is that we do things exactly like OPAM so we can download the same packages as OPAM.

The actual amount of code in dune that's not vendored libraries and their dependencies is fairly minimal.

Leonidas-from-XIV commented on vendor/opam/configure.ac:
```diff
@@ -0,0 +1,494 @@
+dnl The line below must be formatted AC_INIT(opam,VERSION) with no extra spaces
```
Since we now have [ocaml-dune/opam](https://github.com/ocaml-dune/opam) I've put the changes into the `vendored` branch, so it can be (hopefully easily) rebased in the future and basically precomputed this rule, marking it `(mode fallback)`.

Thus I both dropped `configure.ac` and `get_version.ml` from being vendored.

rgrinberg commented:
Some suggestions to shrink the code space:

1. stub out OpamParallel. This won't be needed because we run everything with dune. This will allow us to get rid of ocamlgraph
2. it seems like you're vendoring opam-file-format twice? Once in vendor/ and once in vendor/opam
3. Get rid of all the json related code.

It should be enough after these 3 steps

rgrinberg commented:
Some other general suggestions:

1. We don't need to depend on stdlib-shims anywhere. We require 4.08 and above anyway
2. We can simplify a lot of the build steps by removing the conditional compilation for old versions of OCaml
3. Random files like opam/shell/*.ml aren't needed
4. We should simplify the folder structure a little bit. There's no need for intermediate src directories.
5. We should replace `re` with `dune_re`.

rgrinberg commented:
I notice that some modules depend on cppo. That's all a no-no and we need a way to avoid that.

rgrinberg commented:
I managed to remove ocamlgraph by stubbing out this function:

```
-        let copies () =
-          OpamParallel.map ~jobs:3
-            ~command:(fun (label, local_dirname, _subpath) ->
-                let text = OpamProcess.make_command_text label label in
-                OpamProcess.Job.with_text text @@
-                (try
-                   OpamFilename.cleandir local_dirname;
-                   OpamFilename.copy_dir ~src:tmpdir ~dst:local_dirname;
-                   Done (Up_to_date label)
-                 with OpamSystem.Process_error r ->
-                   Done (Not_available
-                           (Some label, OpamProcess.result_summary r))))
-            dirnames
-        in
+        let copies () = failwith "stubbed out" in
```

It seems wrong anyway. When inside dune, the number of "jobs" isn't fixed.

Leonidas-from-XIV commented:
I'm working on removing CPPO from upstream OPAM, it seems to be quite feasible: https://github.com/ocaml/opam/pull/5498. OPAM still needs to vendor it because Extlib uses it which is used in CUDF and dose3 but as long as we don't use these, we don't need to deal with CPPO.

Leonidas-from-XIV commented:
Ok, so I've removed the remaining empty `sexp` files, switched to `dune_re` (however it is a bit weird our vendored copy is renamed), rebased our OPAM fork to a version that can build the packages that we're vendoring without requiring CPPO.

rgrinberg commented:
> however it is a bit weird our vendored copy is renamed

Note that that this is done to avoid the following problem:

1. We want to write a test that uses ppx_expect and our own libraries
2. ppx_expect uses re itself
3. ppx_expect is built from the opam switch
4. ppx_expect pulls in an `Re` module, while we use `Dune_re` in our internal libraries

Note that we need different names to avoid linking conflicts in step 4. Ideally, we would in fact rename all the toplevel modules we vendor. However, we don't do this because the process is indeed a bit manual and annoying. I hope to make it a bit simpler one day though.

Do we have any other ideas for what code we can eliminate to reduce the size of this PR?

rgrinberg commented on vendor/opam/src/core/dune:
```diff
@@ -0,0 +1,45 @@
+  (name        opam_core)
+  (synopsis    "OCaml Package Manager core internal stdlib")
+  (libraries   dune_re unix bigarray sha uutf
+               (select opamACL.ml from
```
What's this acl stuff? If we don't need it, then let's remove it because we don't support `select` in bootstrapping.

rgrinberg commented on vendor/opam/src/core/gen/gen.ml:
```diff
@@ -0,0 +1,98 @@
+(* trivial cppo replacement with just enough features to do conditional compilation *)
```
You'll need to modify bootstrap to make this work. Moreover, for this to be usable in bootstrap it cannot use use `Re`.

Leonidas-from-XIV replied:
I've removed the whole thing upstream.

rgrinberg commented:
Try adding this library as a dependency to the dune binary. This should help you view the problems with bootstrap in CI.

rgrinberg commented:
One more question: does "source fetching" here include sources from the local file system. E.g. git from a local repo, or rsync. It would be good test these.

rgrinberg commented on vendor/opam/src/core/opamCompat.pp.ml:
```diff
@@ -0,0 +1,70 @@
+module String =
+#if OCAML_VERSION >= (4, 13, 0)
+  String
+#else
```
There's a way to do this without cppo:

```
module String = struct
  include struct
    [@@@ocaml.warning "-32"]

    let exists = (* put your fallback implementation here *)
  end

  include String
end
```

Leonidas-from-XIV replied:
Good point, being worked on on the OPAM side, I will update once that is merged: https://github.com/ocaml/opam/pull/5508

Leonidas-from-XIV replied:
Rebased the `vendored` branch and updated the code here to use that commit.

rgrinberg commented on test/expect-tests/dune_pkg/dune_pkg_unit_tests.ml:
```diff
@@ -0,0 +1,94 @@
+  let content_length = in_channel_length read_end in
+  let content = really_input_string read_end content_length in
+  close_in read_end;
+  content
```
We have ` Io.String_path.read_file`.

rgrinberg commented on test/expect-tests/dune_pkg/dune_pkg_unit_tests.ml:
```diff
@@ -0,0 +1,94 @@
+
+let url ~port ~filename =
+  let localhost = Unix.inet_addr_loopback |> Unix.string_of_inet_addr in
+  Format.asprintf "http://%s:%d/%s" localhost port filename |> OpamUrl.of_string
```
You can just use `sprintf` here.

rgrinberg commented on vendor/opam/src/core/dune:
```diff
@@ -0,0 +1,45 @@
+                       (opam-core.stubs -> opamStubs.win32.ml)
+                       (                -> opamStubs.dummy.ml)))
+  (flags       (:standard
+               (:include ../ocaml-flags-standard.sexp)))
```
`:include` won't work for bootstrapping so we can just drop it. We disable warnings for vendored sources anyway.

rgrinberg commented on vendor/opam/src/core/dune:
```diff
@@ -0,0 +1,45 @@
+    "let value = \"%{read-strings:developer}\""))
+
+(rule
+  (write-file opamVersionInfo.ml
```
we can replace all this version stuff with some hard coded version file. It should never appear anywhere user facing anyway.

rgrinberg replied:
All of the other `rule` stanzas should be removed as well. None of them are usable in bootstrap.

Leonidas-from-XIV commented:
@rgrinberg Just rebased, will squash if once CI is done; no point in retaining the commit history.
